### PR TITLE
cumulative flow states

### PR DESCRIPTION
**Pull request by SouthEndMusic**
Fixes https://github.com/Deltares/Ribasim/issues/1808

### CODE REVIEW

**Comment by SouthEndMusic** [comment]
Finally all tests are green again after the refactor. It was kind of a pain to make all the book-keeping work out (quite a bit changed in that regard).

# Findings

These findings are based on running the HWS model as taken from the regression tests.

## Water balance

Let's start with the best result: for the HWS model the largest absolute water balance error is `8.83354278613524e-10`.

**_This means that per basin per day, less than a μL is not accounted for._**

Note that this balance error is independent of the solver tolerances, the water balance is inherently closed due to the problem formulation. The errors left are just (accumulated) floating point errors.

## Profile

The refactor didn't introduce any glaring performance problems in the code: 

image


## States and Jacobian sparsity

The state vector is still a component vector, with these components:

```
julia> only(getfield(model.integrator.u, :axes))
Axis(tabulated_rating_curve = 1:11, pump = 12:40, outlet = 41:101, user_demand_inflow = 102:127, user_demand_outflow = 128:153, linear_resistance = 154:153, manning_resistance = 154:239, evaporation = 240:376, infiltration = 377:513, integral = 514:522)
```
i.e. sorted by node type. This means that we can find convergence bottlenecks in a more fine-grained way than before.

Considerations for the states:

- Each conservative node and not exactly integrable boundary node has a state. So `FlowBoundary` doesn't need a state, and `UserDemand` needs states for inflow and outflow separately due to the time-dependent return factor.
- Also `precipitation` and `drainage` don't need their own states, as these are exactly integrable (precipitation uses a fixed area).
- `evaporation` and `infiltration` now have separate states, a bit more on that later.

The Jacobian sparsity looks like this:

plot

Which admittedly looks quite cool, but there are now many more states (going from 137 to 522) and more non-zeros in the Jacobian. To give an indication on how the Jacobian sparsity comes to be: say we have the following model section.

model_detail

The state for the manning flow depends on the forcing states of the connected basins, but also the flow states of all nodes connected to these basins. That is because the manning flow depends on the basin levels, which depend on the basin storages, which depend on the cumulative flows over all flow edges connected to the basins.

## Performance

As expected, the performance takes a hit. The runtime of the HWS model:

- On main: `104.484030 seconds (63.46 M allocations: 10.354 GiB, 0.65% gc time, 0.01% compilation time)` (I thought the HWS model was much faster on main?)
- On this branch: `266.583831 seconds (441.16 M allocations: 104.220 GiB, 1.53% gc time)`

I had hoped I could reduce the number of states by clumping `evaporation` and `infliltration` together into a single state. The problem with that is that the contribution of these forcings over a timestep can then not exactly be split into an evaporation and an infiltration contribution. The contribution of this state of combined forcings over a timestep $\Delta t = t_1 - t_0$ would be:

$$
   \Delta V = \text{infiltration} \cdot \int_{t_0}^{t_1} \phi\left(d(u(t')); 0.1\right) \text{d}t' + \text{evaporation} \cdot \int_{t_0}^{t_1} A\left(u(t')\right)\phi\left(d(u(t');0.1\right)\text{d}t'.
$$

Both integrals are unknowns (except when the basin is not in the 'reduction factor regime'), so it cannot be computed exactly how $\Delta V$ is split up into the infiltration and evaporation contributions. If we find the performance more important than this exact split, we could approximate it. _Note that this approximation does not affect the water balance, because we still know exactly how much water is leaving the basin_.

**Comment by visr** [comment]
I made a plot of the "Test Delwaq coupling" artifact "delwaq_map.nc" variable "ribasim_Continuity", which should be as close to 1 as possible, comparing this branch (new) to the current main (old).

image

So that is very encouraging. (note the +1 on the y axis, so 0 is 1)

```python
import pandas as pd
import xarray as xr

ds_old = xr.open_dataset(r"main\delwaq_map.nc")
da_old = ds_old["ribasim_Continuity"]

ds_new = xr.open_dataset(r"integrated-flow-states\delwaq_map.nc")
da_new = ds_new["ribasim_Continuity"]

df_old = da_old.sel(ribasim_nNodes=3).to_dataframe()["ribasim_Continuity"].rename("old")
df_new = da_new.sel(ribasim_nNodes=3).to_dataframe()["ribasim_Continuity"].rename("new")

df = pd.concat([df_old, df_new], axis=1)
df.plot()
```

**Comment by Huite** [comment]
Looks very promising! The robustness + simplicity with regards to the water balances makes me think this is well worth it.

> Which admittedly looks quite cool, but there are now many more states (going from 137 to 522) and more non-zeros in the Jacobian.

You can reduce bandwidth with algorithms like: https://en.wikipedia.org/wiki/Cuthill%E2%80%93McKee_algorithm

I'm not sure this is relevant: maybe the solver machinery already does things like this. Second, given these sizes, I reckon a lot fits almost wholly within CPU caches, so better bandwidth may have no meaningful effects on memory access patterns.

**Comment by SouthEndMusic** [comment]
> You can reduce bandwidth with algorithms like: https://en.wikipedia.org/wiki/Cuthill%E2%80%93McKee_algorithm

@Huite I have dabbled in that before. As far as I'm aware, the goal of reducing the bandwidth is reducing fill-in in the Jacobian factorization. I looked a little bit in what `OrdindaryDiffEq.jl` does there and the factors I found were quite sparse, so there's probably already some hidden cleverness there

File: core/src/solve.jl [hunk 1]
```diff
@@ -75,63 +68,192 @@ function formulate_continuous_control!(du, p, t)::Nothing
...
+)
+    (; current_cumulative_precipitation, current_cumulative_drainage) = basin
+
+    @. current_storage -= u.evaporation
+    @. current_storage -= u.infiltration
+
+    current_cumulative_precipitation = current_cumulative_precipitation[parent(du)]
+    current_cumulative_drainage = current_cumulative_drainage[parent(du)]
+    @. current_storage += current_cumulative_precipitation
+    @. current_storage += current_cumulative_drainage
```
Comment (visr) [inline core/src/solve.jl:173]:
```suggestion
    current_storage .-= u.evaporation
    current_storage .-= u.infiltration

    current_cumulative_precipitation = current_cumulative_precipitation[parent(du)]
    current_cumulative_drainage = current_cumulative_drainage[parent(du)]
    current_storage .+= current_cumulative_precipitation
    current_storage .+= current_cumulative_drainage
```

File: core/src/solve.jl [hunk 2]
```diff
@@ -198,11 +320,11 @@ function formulate_pid_control!(
 
         if !iszero(K_d)
             dlevel_demand = derivative(target[i], t)
-            du_listened_basin_old = du.storage[listened_node_id.idx]
+            dstorage_listened_basin_old = formulate_dstorage(du, p, t, listened_node_id)#du.storage[listened_node_id.idx]
```
Comment (visr) [inline core/src/solve.jl:323]:
```suggestion
            dstorage_listened_basin_old = formulate_dstorage(du, p, t, listened_node_id)
```

File: docs/concept/equations.qmd [hunk 3]
```diff
@@ -3,76 +3,94 @@ title: "Equations"
...
+This is the error integral for PID control, further explained in [PID equations](/reference/node/pid-control.qmd#equations).
 
-$$
-u(l) = \int_{l_0}^l A(\ell)d\ell.
-$$
+## Exactly integrating flows to minimize the number of states
+
+The more states the problem has, the more time it takes to solve it.
+Therefore we want to minimize the number of states.
+Flows do not have to be states when they can be integrated over time exactly because they do not depend on the other states.
```
[the remark this answers is not in the document]
Comment (SouthEndMusic) [reply docs/concept/equations.qmd:66]: I don't understand exactly what you mean here. In this sentence I am not specifically talking about `FlowBoundary` nodes. Or do you mean that flows can only be exactly integrable when they are a boundary condition?

File: core/src/solve.jl (hunk 1 above)
Comment (SouthEndMusic) [reply core/src/solve.jl:173]: Why?

**Comment by SouthEndMusic** [comment]
I fixed the flaky tests, turns out there was a dictionary involved in setting up the edge data in the flow table which messed up the ordering.

File: core/src/util.jl [hunk 4]
```diff
@@ -973,11 +926,135 @@ function OrdinaryDiffEqNonlinearSolve.relax!(
     @. z_tmp = nlsolver.z + dz
     resid_before = residual(z_tmp, integrator, nlsolver, f)
     relax!(dz, nlsolver, integrator, f, linesearch)
-    @. z_tmp = nlsolver.z + dz
+    @. z_tmp = nlsolver.z + 0.95 * dz
```
[the remark this answers is not in the document]
Comment (SouthEndMusic) [reply core/src/util.jl:929]: This was just a little test, we should test it better before we merge it. It was inspired by Huite's comment about relaxation after backtracking

File: core/src/util.jl [hunk 5]
```diff
@@ -973,11 +926,135 @@ function OrdinaryDiffEqNonlinearSolve.relax!(
...
     resid_after = residual(z_tmp, integrator, nlsolver, f)
 
     # If the residual increased due to the relaxation, reject it
     if resid_after > resid_before
         @. dz = dz_tmp
     end
 end
+
+function build_state_vector(p::Parameters)
+    # It is assumed that the horizontal flow states come first
```
[the remark this answers is not in the document]
Comment (SouthEndMusic) [reply core/src/util.jl:939]: Yep, added some explanation

File: core/src/util.jl [hunk 6]
```diff
@@ -973,11 +926,135 @@ function OrdinaryDiffEqNonlinearSolve.relax!(
...
+        user_demand_outflow = zeros(length(p.user_demand.node_id)),
+        linear_resistance = zeros(length(p.linear_resistance.node_id)),
+        manning_resistance = zeros(length(p.manning_resistance.node_id)),
+        evaporation = zeros(length(p.basin.node_id)),
+        infiltration = zeros(length(p.basin.node_id)),
+        integral = zeros(length(p.pid_control.node_id)),
+    )
+end
+
+function set_state_flow_edges(p::Parameters, u0::ComponentVector)::Parameters
```
[the remark this answers is not in the document]
Comment (SouthEndMusic) [reply core/src/util.jl:954]: I'm not sure about the exclamation mark, technically this function is not mutating the inputs, its creating a new modified version of the input `p`

File: core/src/util.jl [hunk 7]
```diff
@@ -973,11 +926,135 @@ function OrdinaryDiffEqNonlinearSolve.relax!(
...
+    ::ComponentVector{A, B, <:Tuple{<:Axis{NT}}};
+    inflow::Bool = true,
+) where {A, B, NT}
+    component_name = if id.type == NodeType.UserDemand
+        inflow ? :user_demand_inflow : :user_demand_outflow
+    else
+        snake_case(Symbol(id.type))
+    end
+
+    for (comp, range) in zip(keys(NT), values(NT))
```
[the remark this answers is not in the document]
Comment (SouthEndMusic) [reply core/src/util.jl:1062]: Can you elaborate?

File: docs/concept/equations.qmd (hunk 3 above)
[the remark this answers is not in the document]
Comment (visr) [reply docs/concept/equations.qmd:66]: Yeah I meant the latter, though that is clear from the following sentence as well, so let's keep it as-is.

**Review by visr — APPROVED** [review]
